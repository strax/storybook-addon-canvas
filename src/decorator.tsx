import React from "react";
import { Fragment } from "react";
import { StoryDecorator, Renderable } from "@storybook/react";

export function decorator(f: StoryDecorator): StoryDecorator {
  return (render, ctx) => {
    return <Fragment>{f(render, ctx)}</Fragment>
  }
}

export const wrap = (Component: React.ComponentType<{}>): StoryDecorator =>
  render => <Component>{render()}</Component>;

export const compose = (...fs: StoryDecorator[]): StoryDecorator =>
  (story, ctx) => fs.reduceRight((acc, f) => f(() => acc, ctx), story() as Renderable);
