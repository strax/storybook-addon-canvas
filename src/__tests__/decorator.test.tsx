import React from "react";
import { mount } from "enzyme";
import { wrap, compose } from "../decorator";

const Wrapper1: React.SFC = props => <div id="wrapper1">{props.children}</div>;
const Wrapper2: React.SFC = props => <div id="wrapper2">{props.children}</div>;
const Wrapper3: React.SFC = props => <div id="wrapper3">{props.children}</div>;

const render = () => <div>wrapped</div>;
const ctx = { kind: "test", story: "test" }

describe("wrap", () => {
  it("returns a decorator that wraps the render function in the given component", () => {
    const rendered = wrap(Wrapper1)(render, ctx);
    expect(mount(rendered as JSX.Element)).toMatchSnapshot();
  });
});

describe("compose", () => {
  it("composes the given decorators from right to left", () => {
    const decorate = compose(wrap(Wrapper1), wrap(Wrapper2), wrap(Wrapper3));
    expect(mount(decorate(render, ctx) as JSX.Element)).toMatchSnapshot();
  });
});
