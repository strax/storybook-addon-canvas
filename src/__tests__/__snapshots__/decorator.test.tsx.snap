// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose composes the given decorators from right to left 1`] = `
<Wrapper1>
  <div
    id="wrapper1"
  >
    <Wrapper2>
      <div
        id="wrapper2"
      >
        <Wrapper3>
          <div
            id="wrapper3"
          >
            <div>
              wrapped
            </div>
          </div>
        </Wrapper3>
      </div>
    </Wrapper2>
  </div>
</Wrapper1>
`;

exports[`wrap returns a decorator that wraps the render function in the given component 1`] = `
<Wrapper1>
  <div
    id="wrapper1"
  >
    <div>
      wrapped
    </div>
  </div>
</Wrapper1>
`;
